docs: `model_selection` module docstrings

mkalimeri opened the pull request:
# Description

Related to #596

Examples added for

- model_selection.TimeGapSplit
- model_selection.GroupTimeSeriesSplit
- model_selection.ClusterFoldValidation (I see that KlusterFoldValidation has been renamed as ClusterFoldValidation)

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] My code follows the style guidelines (ruff)
- [x] I have commented my code, particularly in hard-to-understand areas
- [NA] I have made corresponding changes to the documentation (also to the readme.md)
- [NA] I have added tests that prove my fix is effective or that my feature works
- [NA] I have added tests to check whether the new feature adheres to the sklearn convention
- [x] New and existing unit tests pass locally with my changes

If you feel your PR is ready for a review, ping @FBruzzesi or @koaning.

mkalimeri committed:
Update model_selection.py

FBruzzesi commented on sklego/model_selection.py:
```diff
@@ -84,6 +84,48 @@ class TimeGapSplit:
+    ### Start date     2024-01-01 00:00:00
+    ### End date       2024-01-01 00:00:09
+    ### Period             0 days 00:00:09
+    ### Unique days                     10
```
How is this returning 10 days? 🤔

Edit: Looking into the code, this is fairly random assumption that the number of unique dates are in days.

mkalimeri replied:
Yes, this should not be unique days but maybe timepoints? Shall we leave this as is for now and open another issue to generate additional/different information in the summary? Happy to pick that up!

koaning replied:
Yeah, this is a nit that would be nice to tackle. I would also totally be open to doing that in this PR, less waiting for merges that way.

mkalimeri replied:
Sure, I can do that. I will come back to you with a plan of what information I would include. Updating this would require changes in tests as well, so it will be more than one file. Is that ok?

mkalimeri replied:
Hi, I got to take a look at this today. We can keep the number of unique days in the summary, but as @FBruzzesi mentioned, we can't assume that the number of days is relevant to all use cases. We could keep the field and correct the calculation, to give an idea of the frequency of data, but it might not be relevant if the granularity of the time series is lower. Another idea is to add the granularity of the time series as a field of the summary. What are your thoughts?

FBruzzesi commented on sklego/model_selection.py:
```diff
@@ -84,6 +84,48 @@ class TimeGapSplit:
+    # Create dataset
+    np.random.seed(1)
+    num_rows = 30
+    df = pd.DataFrame(np.random.randn(num_rows, 4)).rename(columns={0: 'c1', 1: 'c2', 2: 'c3', 3: 'c4'})
```
You can pass the column names directly:

```suggestion
    df = pd.DataFrame(np.random.randn(num_rows, 4), columns=["c1", "c2", "c3", "c4"])
```

FBruzzesi reviewed (commented):
Thanks for the PR @mkalimeri!

The only concern I have is related to the part of code in `TimeGapSplit` summary. @koaning WDYT? 
We can keep it as a follow up, but that's certainly not the right assumption to make in this context

mkalimeri committed:
Updated the code based on feedback on first commit

1. Updated code in TimeGapSplit: corrected the calculation of total number of days in the fold and added a 'frequency' field in the summary, to show the frequency of the data
2. Updated TimeGapSplit example: data frequency is now in hours and the parameters are in days.
3. Updated pytest for summary calculation

mkalimeri commented:
I updated the code based on the feedback from the previous review. I kept the summary field 'num of days' and updated the calculation to reflect the actual number of days. I also added a field 'frequency', to inform about the frequency of the data. I updated example/test to reflect these changes. Let me know if you can think of other informative fields!

mkalimeri committed:
Merge branch 'main' into issue596-model_selection

koaning reviewed (approved):
I am approving, partially to trigger the unit tests

FBruzzesi commented:
@koaning Narwhals CI failed after merging this one (see [scikit-lego job](https://github.com/narwhals-dev/narwhals/actions/runs/30612397190/job/91097902617?pr=3826))

```terminal
FAILED tests/test_model_selection/test_timegapsplit.py::test_timegapsplit_summary - AssertionError: DataFrame.iloc[:, 3] (column name="frequency") are different

DataFrame.iloc[:, 3] (column name="frequency") values are different (100.0 %)
[index]: [(0, train), (0, valid), (1, train), (1, valid), (2, train), (2, valid), (3, train), (3, valid), (4, train), (4, valid), (5, train), (5, valid)]
[left]:  [24h, 24h, 24h, 24h, 24h, 24h, 24h, 24h, 24h, 24h, 24h, 24h]
[right]: [D, D, D, D, D, D, D, D, D, D, D, D]
At positional index 0, first diff: 24h != D
```

koaning commented:
GAh! And here I was trying to save time 😅

FBruzzesi commented:
I just had some time to double check what's happening:

1. As we have and use uv.lock, our dependencies are not getting updated. That's _kind of_ an issue as we are not testing with latest
2. In this specific case: we derive the frequency column of `TimeGapSplit.summary()` via `pd.tseries.frequencies.to_offset(delta).freqstr`. In pandas 3.0 the Day offset became a calendar day (a non-fixed duration, DST-aware). An exact `Timedelta` is a fixed duration, so `to_offset` can no longer map it onto `Day` and falls back to hours:

delta | pandas 2.3.3 | pandas 3.0.5
-- | -- | --
1 day | D | 24h
2 days | 2D | 48h
7 days | 7D | 168h
36 hours | 36h | 36h
90 minutes | 90min | 90min

3. If I try to first bump all version in the uv lock `uv sync --all-extras --all-groups --upgrade`, we get a bunch more tests failing.
4. I can try to make some time next week to fix them :)
